src/transformStateWithClones.js

```diff
@@ -7,7 +7,28 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateArray = [];
+  let newState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        newState = {};
+        break;
+
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+    }
+    stateArray.push({ ...newState });
```

Comment:
The cloning of 'newState' into the 'stateArray' is done inside the loop after each action. This is correct, but it's important to note that if an action type is not recognized (which is currently not handled), it would still push the current state into the array. This could be avoided by handling the default case in the switch statement.